components/profile.js

```diff
@@ -103,15 +104,283 @@ SteamCommunity.prototype.editProfile = function(settings, callback) {
+									break;
+
+								case 'game':
+									values["profile_showcase[6]"] = 6;
+
+									if (settings[i][type].hasOwnProperty("values")) {
+										values["rgShowcaseConfig[6][0][appid]"] = settings[i][type]["values"];
+									}
+									break;
+
+								case 'badge':
+									values["profile_showcase[5]"] = 5;
+
+									if (settings[i][type].hasOwnProperty("values")) {
+
+										if (settings[i][type]["values"].hasOwnProperty("style")) {
+											var styles = ["rare", "selected", null, "recent", "random"];
+											values["profile_showcase_style_5"] = styles.indexOf(settings[i][type]["values"]["style"]);
+										}
+
+										if (settings[i][type]["values"].hasOwnProperty("badges")) {
+											var types = ["badgeid", "appid", "border_color"];
+											for (var n = 0; n < 6; n++) {
+												for (var t in type){
+													if (settings[i][type]["values"]["badges"][n] != undefined) {
+														if (settings[i][type]["values"]["badges"][n].hasOwnProperty(types[t])){
+															values["rgShowcaseConfig[5][" + n + "][" + types[t] + "]"] = settings[i][type]["values"]["badges"][n][types[t]] || values["rgShowcaseConfig[5][" + n + "][" + types[t] + "]"] || "";
+														}
+													}
+												}
+											}
+										}
+									}
+
+									break;
+
+								case 'rareachievements':
+									values["profile_showcase[1]"] = 1;
+									break;
+
+								case 'screenshot':
+									values["profile_showcase[7]"] = 7;
+
+									if (settings[i][type].hasOwnProperty("values")) {
+										for (var n = 0; n < 4; n++) {
+											if (settings[i][type]["values"][n] != undefined) {
+												values["rgShowcaseConfig[7][" + n + "][publishedfileid]"] = settings[i][type]["values"][n];
+											}
+										}
+									}
+									break;
+
+								case 'group':
+									values["profile_showcase[9]"] = 9;
+
+									if (settings[i][type].hasOwnProperty("values")) {
+										if (typeof settings[i][type]["values"] === 'object' && settings[i][type]["values"].getSteamID64) {
+											values["rgShowcaseConfig[9][0][accountid]"] = settings[i][type]["values"].getSteamID64();
+										} else {
+											values["rgShowcaseConfig[9][0][accountid]"] = new SteamID(settings[i][type]["values"]).getSteamID64();
+										}
+									}
+									break;
+
+								case 'review':
+									values["profile_showcase[10]"] = 10;
+
+									if (settings[i][type].hasOwnProperty("values")) {
+										values["rgShowcaseConfig[10][0][appid]"] = settings[i][type]["values"];
+									}
+									break;
+
+								case 'workshop':
+									values["profile_showcase[11]"] = 11;
+
+									if (settings[i][type].hasOwnProperty("values")) {
+										values["rgShowcaseConfig[11][0][appid]"] = settings[i][type]["values"]["appid"];
+										values["rgShowcaseConfig[11][0][publishedfileid]"] = settings[i][type]["values"]["publishedfileid"];
+									}
+									break;
+
+								case 'guide':
```

Comment:
A lot of these cases contain very similar code. It's hard for me to believe this can't be done in a more efficient way that would be more easily transferable to more showcases as Valve adds them.

Reply:
There's always a "proper" way to do something, and I think doing it in a way which allows more showcases to be easily added on is a must. I guess the current method isn't too bad, especially after a little cleanup.

I guess in the ideal world there would be a Showcases class containing the data that every showcase might want, such as `item`, `title`, `badge`, etc. Then the code here would be able to get passed a specific showcase class from the user and add that to their profile. Maybe this is a little too ambitious for something that probably won't be used too often, but it feels like there's still some work that could be done to make this all more friendly.